scipy/optimize/_qap.py

```diff
@@ -0,0 +1,784 @@
+
+    Returns
+    -------
+    res : OptimizeResult
+        A :class:`scipy.optimize.OptimizeResult` containing the following
+        fields.
+
+        col_ind : 1-D array
+            An array of column indices corresponding with the best
+            permutation of the nodes of `B` found.
+        fun : float
+            The corresponding value of the objective function.
+        nit : int
+            The number of Franke-Wolfe iterations performed.
+
+    Notes
+    -----
+    The algorithm may be sensitive to the initial permutation matrix (or
+    search "position") due to the possibility of several local minima
+    within the feasible region. A barycenter initialization is more likely to
+    result in a better solution than a single random initialization. However,
+    ``quadratic_assignment`` calling several times with different random
+    initializations may result in a better optimum at the cost of longer
+    total execution time.
+
+    Examples
+    --------
+    As mentioned above, a barycenter initialization often results in a better
+    solution than a single random initialization.
+
+    >>> np.random.seed(0)
+    >>> n = 15
+    >>> A = np.random.rand(n, n)
+    >>> B = np.random.rand(n, n)
+    >>> res = quadratic_assignment(A, B)  # FAQ is default method
+    >>> print(res.fun)
+    46.871483385480545 # may vary
+
+    >>> options = {"init_weight": 0}  # use 100% random initialization
+    >>> res = quadratic_assignment(A, B, options=options)
+    >>> print(res.fun)
+    46.91558492779956 # may vary
+
+    However, consider running from several random initializations and keeping
+    the best result.
+    >>> res = min([quadratic_assignment(A, B, options=options)
+    ...            for i in range(30)], key=lambda x: x.fun)
+    >>> print(res.fun)
+    46.50791824173632 # may vary
+
+    The '2-opt' method can be used to further refine the results.
+    >>> options = {"partial_guess": np.array([np.arange(n), res.col_ind]).T}
+    >>> res = quadratic_assignment(A, B, method="2opt", options=options)
+    >>> print(res.fun)
+    46.47160735721583 # may vary
+
+    References
+    ----------
+    .. [1] J.T. Vogelstein, J.M. Conroy, V. Lyzinski, L.J. Podrazik,
+           S.G. Kratzer, E.T. Harley, D.E. Fishkind, R.J. Vogelstein, and
+           C.E. Priebe, "Fast approximate quadratic programming for graph
+           matching," PLOS one, vol. 10, no. 4, p. e0121002, 2015,
+           https://doi.org/10.1371/journal.pone.0121002
+
+    .. [2] D. Fishkind, S. Adali, H. Patsolic, L. Meng, D. Singh, V. Lyzinski,
+           C. Priebe, "Seeded graph matching", Pattern Recognit. 87 (2019):
+           203-215, https://doi.org/10.1016/j.patcog.2018.09.014
+    """
+
+    _check_unknown_options(unknown_options)
+
+    maxiter = operator.index(maxiter)
+
+    # ValueError check
+    A, B, partial_match = _common_input_validation(A, B, partial_match)
+
+    msg = None
+    if (A < 0).any() or (B < 0).any():
```

Comment:
Is this a real restriction of the algorithm?

Reply:
This is not a "real" restriction for the algorithm but I had originally added it to help the user not have unexpected things happen. For example, if one of the matrices has a mixture of positive and negative values, this is likely a user mistake. But, that being said, I wouldn't be against removing this check, since having negative adjacency matrices wouldn't cause the code to break.

Reply:
I would remove the restriction.